extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java

```diff
@@ -45,19 +49,43 @@ public void setup(OidcConfig config, RuntimeValue<Vertx> vertx, BeanContainer be
                     .setPublicKey(config.publicKey.get()));
         }
 
-        CompletableFuture<OAuth2Auth> cf = new CompletableFuture<>();
-        KeycloakAuth.discover(vertx.getValue(), options, new Handler<AsyncResult<OAuth2Auth>>() {
-            @Override
-            public void handle(AsyncResult<OAuth2Auth> event) {
-                if (event.failed()) {
-                    cf.completeExceptionally(toOidcException(event.cause()));
+        final long connectionDelayInSecs = config.connectionDelay.isPresent() ? config.connectionDelay.get().toMillis() / 1000
+                : 0;
+        final long connectionRetryCount = connectionDelayInSecs > 1 ? connectionDelayInSecs / 2 : 1;
+        if (connectionRetryCount > 1) {
+            LOG.infof("Connecting to IDP for up to %d times every 2 seconds", connectionRetryCount);
+        }
+
+        OAuth2Auth auth = null;
+        for (long i = 0; i < connectionRetryCount; i++) {
+            try {
+                CompletableFuture<OAuth2Auth> cf = new CompletableFuture<>();
+                KeycloakAuth.discover(vertx.getValue(), options, new Handler<AsyncResult<OAuth2Auth>>() {
+                    @Override
+                    public void handle(AsyncResult<OAuth2Auth> event) {
+                        if (event.failed()) {
+                            cf.completeExceptionally(toOidcException(event.cause()));
+                        } else {
+                            cf.complete(event.result());
+                        }
+                    }
+                });
+
+                auth = cf.join();
+                break;
+            } catch (OIDCException ex) {
+                if (i + 1 < connectionRetryCount) {
+                    try {
+                        Thread.sleep(2000);
```

Comment:
I wonder if we actually want a more generic approach to dealing with this, as there are quite a few different considerations here:
- Should we use a similar approach for all resources (database connections etc). Would some kind of unified 'resource timeout' parameter make more sense
- How does this affect liveliness/readiness probes. AFAIK if we are trying to connect to the OIDC server then we should be live but not ready
- What happens if the OIDC server goes down while we are running?

This is probably actually a good topic for the F2F.